go/teams/seitan_v2.go

```diff
@@ -0,0 +1,276 @@
+)
+
+// Seitan tokens v2 have a '+' as the sixth character. We use this
+// to distinguish from email invite tokens (and team names).
+// See `IsSeitanyV2`
+const seitanEncodedIKeyV2PlusOffset = 6
+
+// "Invite Key Version 2"
+type SeitanIKeyV2 string
+
+// "Packed Encrypted Invite Key Version 2"
+// All following 3 structs should be considered one. When any changes,
+// Version in PEPubKey has to be bumped up.
+type SeitanPEPubKey struct {
+	_struct              bool `codec:",toarray"`
+	Version              uint
+	TeamKeyGeneration    keybase1.PerTeamKeyGeneration
+	RandomNonce          keybase1.BoxNonce
+	EncryptedKeyAndLabel []byte // keybase1.SeitanPubKeyAndLabel MsgPacked and encrypted
+}
+
+func GenerateIKeyV2() (ikey SeitanIKeyV2, err error) {
+	buf, err := generateIKey(seitanEncodedIKeyV2PlusOffset)
+	if err != nil {
+		return ikey, err
+	}
+	return SeitanIKeyV2(string(buf)), err
+}
+
+// IsSeitany is a very conservative check of whether a given string looks
+// like a Seitan token. We want to err on the side of considering strings
+// Seitan tokens, since we don't mistakenly want to send botched Seitan
+// tokens to the server.
+func IsSeitanyV2(s string) bool {
+	return len(s) > seitanEncodedIKeyV2PlusOffset && strings.IndexByte(s, '+') > 1
+}
+
+// GenerateIKeyV2FromString safely creates SeitanIKey value from
+// plaintext string. Only format is checked - any 18-character token
+// with '+' character at position 6 can be "Invite Key". Alphabet is
+// not checked, as it is only a hint for token generation and it can
+// change over time, but we assume that token length stays the same.
+func GenerateIKeyV2FromString(token string) (ikey SeitanIKeyV2, err error) {
+	if len(token) != SeitanEncodedIKeyLength {
+		return ikey, fmt.Errorf("invalid token length: expected %d characters, got %d", SeitanEncodedIKeyLength, len(token))
+	}
+	if token[seitanEncodedIKeyV2PlusOffset] != '+' {
+		return ikey, fmt.Errorf("invalid token format: expected %dth character to be '+'", seitanEncodedIKeyV2PlusOffset+1)
+	}
+
+	return SeitanIKeyV2(strings.ToLower(token)), nil
+}
+
+func (ikey SeitanIKeyV2) String() string {
+	return strings.ToLower(string(ikey))
+}
+
+// "Stretched Invite Key"
+type SeitanSIKeyV2 [SeitanScryptKeylen]byte
+
+func (ikey SeitanIKeyV2) GenerateSIKey() (sikey SeitanSIKeyV2, err error) {
+	sikeyv1, err := SeitanIKey(ikey).GenerateSIKey()
+	return SeitanSIKeyV2(sikeyv1), err
+}
+
+func (sikey SeitanSIKeyV2) GenerateTeamInviteID() (id SCTeamInviteID, err error) {
+	type InviteStagePayload struct {
+		Stage   string `codec:"stage" json:"stage"`
+		Version int    `codec:"version" json:"version"`
+	}
+
+	payload, err := libkb.MsgpackEncode(InviteStagePayload{
+		Stage:   "invite_id",
+		Version: 2,
+	})
+	if err != nil {
+		return id, err
+	}
+	return generateTeamInviteID(sikey[:], payload)
+}
+
+func (sikey SeitanSIKeyV2) generateKeyPair() (key libkb.NaclSigningKeyPair, err error) {
```

Comment:
I know this is arbitrary at this point - since both consumer and api live in one package - but maybe this one could be "public" and start with uppercase, because whoever wants to generate seitan should call it. LMK if I got that one wrong though!

Reply:
i want to leave this private -- i use it in the tests but the public should use `GenerateSignature`

Reply:
ah if it's test only then sure, sorry